### PR TITLE
Persistence Feature

### PR DESCRIPTION
**Pull request by imesh94**
## Purpose
Resolves: #18

## Goals
Adding the capability to store Test Configuration data, Report data and, User data in a database.

## Approach
An embeded database was used to store data.
Uses H2 Database Engine.

## Automation tests
 - Unit tests 
   > Code coverage information
 - Integration tests
   > Details about the test cases and coverage

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes/no
 - Ran FindSecurityBugs plugin and verified report? yes/no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes/no

### CODE REVIEW

File: components/org.wso2.finance.open.banking.conformance.mgt/src/main/java/org/wso2/finance/open/banking/conformance/mgt/dao/impl/ReportDAOImpl.java [hunk 1]
```diff
@@ -0,0 +1,277 @@
...
+        Gson gson = new Gson();
+        Report report = null;
+        Connection conn = DBConnector.getConnection();
+        PreparedStatement stmt = null;
+
+        try {
+            String sql = SQLConstants.RETRIEVE_REPORT;
+            stmt = conn.prepareStatement(sql);
+            stmt.setInt(1, reportID);
+            ResultSet rs = stmt.executeQuery();
```
Comment (charithadeshapriya) [inline components/org.wso2.finance.open.banking.conformance.mgt/src/main/java/org/wso2/finance/open/banking/conformance/mgt/dao/impl/ReportDAOImpl.java]: This resource is not properly closed on the checked exception. close it properly in the
catch(SQLException se) {
}

File: components/org.wso2.finance.open.banking.conformance.mgt/src/main/java/org/wso2/finance/open/banking/conformance/mgt/dao/impl/ReportDAOImpl.java [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+    public List<Report> getReports(int testID) {
+
+        Gson gson = new Gson();
+        List<Report> reports = new ArrayList<Report>();
+        Connection conn = DBConnector.getConnection();
+        PreparedStatement stmt = null;
+
+        try {
+            String sql = SQLConstants.DELETE_EMPTY_REPORTS;
+            stmt = conn.prepareStatement(sql);
```
Comment (charithadeshapriya) [inline components/org.wso2.finance.open.banking.conformance.mgt/src/main/java/org/wso2/finance/open/banking/conformance/mgt/dao/impl/ReportDAOImpl.java]: This resource is not properly closed on the checked exception. close it properly in the
catch(SQLException se) {
}

File: components/org.wso2.finance.open.banking.conformance.mgt/src/main/java/org/wso2/finance/open/banking/conformance/mgt/dao/impl/ReportDAOImpl.java [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+            stmt.executeUpdate();
+            sql = SQLConstants.RETRIEVE_REPORTS;
+            stmt = conn.prepareStatement(sql);
+            stmt.setInt(1, testID);
+            ResultSet rs = stmt.executeQuery();
```
Comment (charithadeshapriya) [inline components/org.wso2.finance.open.banking.conformance.mgt/src/main/java/org/wso2/finance/open/banking/conformance/mgt/dao/impl/ReportDAOImpl.java]: This resource is not properly closed on the checked exception. close it properly in the
catch(SQLException se) {
}

File: components/org.wso2.finance.open.banking.conformance.mgt/src/main/java/org/wso2/finance/open/banking/conformance/mgt/dao/impl/ReportDAOImpl.java [hunk 4]
```diff
@@ -0,0 +1,277 @@
...
+import java.sql.*;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Implementation of the ReportDAO
+ * Responsible for the database operations regarding reports
+ */
+public class ReportDAOImpl implements ReportDAO {
+
```
Comment (charithadeshapriya) [inline components/org.wso2.finance.open.banking.conformance.mgt/src/main/java/org/wso2/finance/open/banking/conformance/mgt/dao/impl/ReportDAOImpl.java]: In general the cyclomatic complexity of following methods seems to be high and see if there is a way to refactor these methods and make them simple.
storeReport
getNewReportID
getReport
getReports

File: components/org.wso2.finance.open.banking.conformance.mgt/src/main/java/org/wso2/finance/open/banking/conformance/mgt/dao/impl/TestPlanDAOImpl.java [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+
+        java.util.Date dt = new java.util.Date();
+        java.text.SimpleDateFormat sdf = new java.text.SimpleDateFormat("yyyy-MM-dd HH:mm:ss");
+        String currentTime = sdf.format(dt);
+        String testPlanJson = gson.toJson(testPlan);
+        PreparedStatement stmt = null;
+        String sql;
+        try {
+            sql = SQLConstants.CREATE_TESTPLAN;
+            stmt = conn.prepareStatement(sql, Statement.RETURN_GENERATED_KEYS);
```
Comment (charithadeshapriya) [inline components/org.wso2.finance.open.banking.conformance.mgt/src/main/java/org/wso2/finance/open/banking/conformance/mgt/dao/impl/TestPlanDAOImpl.java]: resource is not closed properly

File: components/org.wso2.finance.open.banking.conformance.mgt/src/main/java/org/wso2/finance/open/banking/conformance/mgt/dao/impl/TestPlanDAOImpl.java [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+            stmt.setString(1, userID);
+            stmt.setString(2, testPlanJson);
+            stmt.setString(3, currentTime);
+            stmt.executeUpdate();
+
+            ResultSet rs = stmt.getGeneratedKeys();
```
Comment (charithadeshapriya) [inline components/org.wso2.finance.open.banking.conformance.mgt/src/main/java/org/wso2/finance/open/banking/conformance/mgt/dao/impl/TestPlanDAOImpl.java]: resource is not closed properly

File: components/org.wso2.finance.open.banking.conformance.mgt/src/main/java/org/wso2/finance/open/banking/conformance/mgt/dao/impl/TestPlanDAOImpl.java [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+        Gson gson = new Gson();
+        TestPlan testPlan = new TestPlan();
+        Connection conn = DBConnector.getConnection();
+        PreparedStatement stmt = null;
+
+        try {
+            String sql =  SQLConstants.RETRIEVE_TESTPLAN;
+            stmt = conn.prepareStatement(sql);
+            stmt.setInt(1, testID);
+            ResultSet rs = stmt.executeQuery();
```
Comment (charithadeshapriya) [inline components/org.wso2.finance.open.banking.conformance.mgt/src/main/java/org/wso2/finance/open/banking/conformance/mgt/dao/impl/TestPlanDAOImpl.java]: resource is not closed properly

File: components/org.wso2.finance.open.banking.conformance.mgt/src/main/java/org/wso2/finance/open/banking/conformance/mgt/dao/impl/TestPlanDAOImpl.java [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+        ReportDAO reportDAO = new ReportDAOImpl();
+        List<Report> reports;
+        PreparedStatement stmt = null;
+
+        try {
+            // Execute query
+            String sql =  SQLConstants.RETRIEVE_TESTPLANS;
+            stmt = conn.prepareStatement(sql);
+            stmt.setString(1, userID);
+            ResultSet rs = stmt.executeQuery();
```
Comment (charithadeshapriya) [inline components/org.wso2.finance.open.banking.conformance.mgt/src/main/java/org/wso2/finance/open/banking/conformance/mgt/dao/impl/TestPlanDAOImpl.java]: resource is not closed properly

File: components/org.wso2.finance.open.banking.conformance.mgt/src/main/java/org/wso2/finance/open/banking/conformance/mgt/dao/impl/TestPlanDAOImpl.java [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+            conn.close();
+        } catch(SQLException se) {
+            se.printStackTrace();
+        } catch(Exception e) {
+            e.printStackTrace();
+        } finally {
+            try{
+                if(stmt!=null) stmt.close();
+            } catch(SQLException se2) {
+            }
```
Comment (charithadeshapriya) [inline components/org.wso2.finance.open.banking.conformance.mgt/src/main/java/org/wso2/finance/open/banking/conformance/mgt/dao/impl/TestPlanDAOImpl.java]: fill above block of code

File: components/org.wso2.finance.open.banking.conformance.mgt/src/main/java/org/wso2/finance/open/banking/conformance/mgt/dao/impl/TestPlanDAOImpl.java [hunk 10]
```diff
@@ -0,0 +1,190 @@
...
```
Comment (charithadeshapriya) [inline components/org.wso2.finance.open.banking.conformance.mgt/src/main/java/org/wso2/finance/open/banking/conformance/mgt/dao/impl/TestPlanDAOImpl.java]: fill above block of code

File: components/org.wso2.finance.open.banking.conformance.mgt/src/main/java/org/wso2/finance/open/banking/conformance/mgt/dao/impl/TestPlanDAOImpl.java [hunk 11]
```diff
@@ -0,0 +1,190 @@
...
```
Comment (charithadeshapriya) [inline components/org.wso2.finance.open.banking.conformance.mgt/src/main/java/org/wso2/finance/open/banking/conformance/mgt/dao/impl/TestPlanDAOImpl.java]: fill above block of code

File: components/org.wso2.finance.open.banking.conformance.mgt/src/main/java/org/wso2/finance/open/banking/conformance/mgt/dao/impl/TestPlanDAOImpl.java [hunk 12]
```diff
@@ -0,0 +1,190 @@
...
+
+import java.sql.*;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Implementation of the TestPlanDAO
+ * Responsible for the database operations regarding test-plans
+ */
```
Comment (charithadeshapriya) [inline components/org.wso2.finance.open.banking.conformance.mgt/src/main/java/org/wso2/finance/open/banking/conformance/mgt/dao/impl/TestPlanDAOImpl.java]: In general cyclomatic complexity of the methods seems high and please check if methods can be further broken down and make them simple.

File: components/org.wso2.finance.open.banking.conformance.mgt/src/main/java/org/wso2/finance/open/banking/conformance/mgt/dao/impl/UserDAOImpl.java [hunk 13]
```diff
@@ -0,0 +1,188 @@
...
+
+package org.wso2.finance.open.banking.conformance.mgt.dao.impl;
+
+import org.wso2.finance.open.banking.conformance.mgt.dao.UserDAO;
+import org.wso2.finance.open.banking.conformance.mgt.db.DBConnector;
+import org.wso2.finance.open.banking.conformance.mgt.db.SQLConstants;
+import org.wso2.finance.open.banking.conformance.mgt.dto.UserDTO;
+
+import java.sql.*;
+
```
Comment (charithadeshapriya) [inline components/org.wso2.finance.open.banking.conformance.mgt/src/main/java/org/wso2/finance/open/banking/conformance/mgt/dao/impl/UserDAOImpl.java]: same empty code blocks and improper resource closing issues can be seen here too.

File: components/org.wso2.finance.open.banking.conformance.mgt/src/main/java/org/wso2/finance/open/banking/conformance/mgt/db/SQLConstants.java [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+ */
+
+/**
+ * This class will hold all SQL queries needed for the
+ * application as constants.
+ */
+package org.wso2.finance.open.banking.conformance.mgt.db;
+
+public class SQLConstants {
+    /* Create Tables */
```
Comment (charithadeshapriya) [inline components/org.wso2.finance.open.banking.conformance.mgt/src/main/java/org/wso2/finance/open/banking/conformance/mgt/db/SQLConstants.java]: add a private constructor to hide the implicit public one

File: components/org.wso2.finance.open.banking.conformance.mgt/src/main/java/org/wso2/finance/open/banking/conformance/mgt/dto/UserDTO.java [hunk 15]
```diff
@@ -0,0 +1,80 @@
...
+            BigInteger no = new BigInteger(1, messageDigest);
+            String hashtext = no.toString(16);
+
+            while (hashtext.length() < 32) {
+                hashtext = "0" + hashtext;
+            }
+            return hashtext;
+        }
+        catch (NoSuchAlgorithmException e) {
+            throw new RuntimeException(e);
```
Comment (charithadeshapriya) [inline components/org.wso2.finance.open.banking.conformance.mgt/src/main/java/org/wso2/finance/open/banking/conformance/mgt/dto/UserDTO.java]: define and throw a dedicated exception rather than trowing a generic one

File: components/org.wso2.finance.open.banking.conformance.api/src/main/java/org/wso2/finance/open/banking/conformance/api/services/ResultsAPI.java [hunk 16]
```diff
@@ -45,13 +50,17 @@
...
-        return new TestResultDTO(
-                this.runnerManager.getTestPlan(testId),
-                this.runnerManager.getReport(testId, reportId)
-        );
+        TestPlanDAO testPlanDAO = new TestPlanDAOImpl();
+        ReportDAO reportDAO = new ReportDAOImpl();
+        Report report = reportDAO.getReport(reportId);
+
+        if(report == null){
+            report =  this.runnerManager.getReport(testId, reportId);
```
[the remark this answers is not in the document]
Comment (imesh94) [reply components/org.wso2.finance.open.banking.conformance.api/src/main/java/org/wso2/finance/open/banking/conformance/api/services/ResultsAPI.java]: done

File: components/org.wso2.finance.open.banking.conformance.mgt/src/main/java/org/wso2/finance/open/banking/conformance/mgt/dao/impl/ReportDAOImpl.java [hunk 17]
```diff
@@ -0,0 +1,277 @@
...
+            conn.close();
+        } catch(SQLException se) {
+            se.printStackTrace();
+        } catch(Exception e) {
+            e.printStackTrace();
+        } finally {
+            try{
+                if(stmt!=null) stmt.close();
+            } catch(SQLException se2) {
+            }
```
[the remark this answers is not in the document]
Comment (imesh94) [reply components/org.wso2.finance.open.banking.conformance.mgt/src/main/java/org/wso2/finance/open/banking/conformance/mgt/dao/impl/ReportDAOImpl.java]: done

File: components/org.wso2.finance.open.banking.conformance.mgt/src/main/java/org/wso2/finance/open/banking/conformance/mgt/dao/impl/ReportDAOImpl.java [hunk 18]
```diff
@@ -0,0 +1,277 @@
...
+
+        java.util.Date dt = new java.util.Date();
+        java.text.SimpleDateFormat sdf = new java.text.SimpleDateFormat("yyyy-MM-dd HH:mm:ss");
+        String currentTime = sdf.format(dt);
+
+        String reportJson = gson.toJson(report);
+        PreparedStatement stmt = null;
+        try {
+            String sql = SQLConstants.CREATE_REPORT;
+            stmt = conn.prepareStatement(sql);
```
[the remark this answers is not in the document]
Comment (imesh94) [reply components/org.wso2.finance.open.banking.conformance.mgt/src/main/java/org/wso2/finance/open/banking/conformance/mgt/dao/impl/ReportDAOImpl.java]: done

File: components/org.wso2.finance.open.banking.conformance.mgt/src/main/java/org/wso2/finance/open/banking/conformance/mgt/dao/impl/ReportDAOImpl.java (hunk 3 above)
Comment (imesh94) [reply components/org.wso2.finance.open.banking.conformance.mgt/src/main/java/org/wso2/finance/open/banking/conformance/mgt/dao/impl/ReportDAOImpl.java]: done

File: components/org.wso2.finance.open.banking.conformance.mgt/src/main/java/org/wso2/finance/open/banking/conformance/mgt/db/DBConnector.java [hunk 19]
```diff
@@ -0,0 +1,67 @@
...
+ */
+public class DBConnector {
+    private static BasicDataSource dataSource;
+    private static final String DB_URL = "jdbc:h2:~/obsuit-reports";
+    private static final String JDBC_DRIVER = "org.h2.Driver";
+
+    //  Database credentials
+    private static final String USER = "obuser";
+    private static final String PASS = "obpass";
+
```
[the remark this answers is not in the document]
Comment (imesh94) [reply components/org.wso2.finance.open.banking.conformance.mgt/src/main/java/org/wso2/finance/open/banking/conformance/mgt/db/DBConnector.java]: done

File: components/org.wso2.finance.open.banking.conformance.mgt/src/main/java/org/wso2/finance/open/banking/conformance/mgt/db/SQLConstants.java (hunk 14 above)
Comment (imesh94) [reply components/org.wso2.finance.open.banking.conformance.mgt/src/main/java/org/wso2/finance/open/banking/conformance/mgt/db/SQLConstants.java]: done

File: components/org.wso2.finance.open.banking.conformance.mgt/src/main/java/org/wso2/finance/open/banking/conformance/mgt/dto/UserDTO.java (hunk 15 above)
Comment (imesh94) [reply components/org.wso2.finance.open.banking.conformance.mgt/src/main/java/org/wso2/finance/open/banking/conformance/mgt/dto/UserDTO.java]: done

File: components/org.wso2.finance.open.banking.conformance.mgt/src/main/java/org/wso2/finance/open/banking/conformance/mgt/dao/impl/ReportDAOImpl.java [hunk 20]
```diff
@@ -0,0 +1,277 @@
...
+            conn.close();
+        } catch(SQLException se) {
+            se.printStackTrace();
+        } catch(Exception e) {
+            e.printStackTrace();
+        } finally {
+            try{
+                if(stmt!=null) stmt.close();
+            } catch(SQLException se2) {
+            }
```
Comment (Ashi1993) [inline components/org.wso2.finance.open.banking.conformance.mgt/src/main/java/org/wso2/finance/open/banking/conformance/mgt/dao/impl/ReportDAOImpl.java]: Fill this block of code

File: components/org.wso2.finance.open.banking.conformance.api/src/main/java/org/wso2/finance/open/banking/conformance/api/services/ResultsAPI.java (hunk 16 above)
[the remark this answers is not in the document]
Comment (udarakr) [reply components/org.wso2.finance.open.banking.conformance.api/src/main/java/org/wso2/finance/open/banking/conformance/api/services/ResultsAPI.java]: what are these testId and reportId? are these human-readable or an UUID?

File: components/org.wso2.finance.open.banking.conformance.mgt/pom.xml [hunk 21]
```diff
@@ -39,6 +39,18 @@
             <artifactId>gson</artifactId>
             <version>${com.google.code.gson.version}</version>
         </dependency>
+
+        <dependency>
+            <groupId>com.h2database</groupId>
+            <artifactId>h2</artifactId>
+            <version>1.4.196</version>
```
Comment (udarakr) [inline components/org.wso2.finance.open.banking.conformance.mgt/pom.xml]: take these version values to the root pom

File: components/org.wso2.finance.open.banking.conformance.mgt/src/main/java/org/wso2/finance/open/banking/conformance/mgt/dao/ReportDAO.java [hunk 22]
```diff
@@ -0,0 +1,16 @@
+package org.wso2.finance.open.banking.conformance.mgt.dao;
```
Comment (udarakr) [inline components/org.wso2.finance.open.banking.conformance.mgt/src/main/java/org/wso2/finance/open/banking/conformance/mgt/dao/ReportDAO.java]: add license header to all the files(there are multiple files without the header)

File: components/org.wso2.finance.open.banking.conformance.mgt/src/main/java/org/wso2/finance/open/banking/conformance/mgt/dao/impl/ReportDAOImpl.java [hunk 23]
```diff
@@ -0,0 +1,277 @@
...
+                stmt = conn.prepareStatement(sql);
+                stmt.setString(1, reportJson);
+                stmt.setString(2, currentTime);
+                stmt.setInt(3, generatedReportID);
+                stmt.executeUpdate();
+            }
+            stmt.close();
+            conn.close();
+        } catch(SQLException se) {
+            se.printStackTrace();
```
Comment (udarakr) [inline components/org.wso2.finance.open.banking.conformance.mgt/src/main/java/org/wso2/finance/open/banking/conformance/mgt/dao/impl/ReportDAOImpl.java]: use proper messages with proper log levels

File: components/org.wso2.finance.open.banking.conformance.api/src/main/java/org/wso2/finance/open/banking/conformance/api/services/ResultsAPI.java (hunk 16 above)
[the remark this answers is not in the document]
Comment (imesh94) [reply components/org.wso2.finance.open.banking.conformance.api/src/main/java/org/wso2/finance/open/banking/conformance/api/services/ResultsAPI.java]: Those are human readable. Primary key fields of the TestPlan and Report tables.

File: components/org.wso2.finance.open.banking.conformance.mgt/pom.xml (hunk 21 above)
Comment (imesh94) [reply components/org.wso2.finance.open.banking.conformance.mgt/pom.xml]: done

File: components/org.wso2.finance.open.banking.conformance.mgt/src/main/java/org/wso2/finance/open/banking/conformance/mgt/dao/ReportDAO.java (hunk 22 above)
Comment (imesh94) [reply components/org.wso2.finance.open.banking.conformance.mgt/src/main/java/org/wso2/finance/open/banking/conformance/mgt/dao/ReportDAO.java]: done

File: components/org.wso2.finance.open.banking.conformance.mgt/src/main/java/org/wso2/finance/open/banking/conformance/mgt/dao/impl/ReportDAOImpl.java (hunk 20 above)
Comment (imesh94) [reply components/org.wso2.finance.open.banking.conformance.mgt/src/main/java/org/wso2/finance/open/banking/conformance/mgt/dao/impl/ReportDAOImpl.java]: done

File: components/org.wso2.finance.open.banking.conformance.api/src/main/java/org/wso2/finance/open/banking/conformance/api/services/ResultsAPI.java (hunk 16 above)
[the remark this answers is not in the document]
Comment (Ashi1993) [reply components/org.wso2.finance.open.banking.conformance.api/src/main/java/org/wso2/finance/open/banking/conformance/api/services/ResultsAPI.java]: Shall we add a comment there, so that future developers will understand?

File: components/org.wso2.finance.open.banking.conformance.mgt/src/main/java/org/wso2/finance/open/banking/conformance/mgt/dao/impl/ReportDAOImpl.java (hunk 4 above)
Comment (imesh94) [reply components/org.wso2.finance.open.banking.conformance.mgt/src/main/java/org/wso2/finance/open/banking/conformance/mgt/dao/impl/ReportDAOImpl.java]: done. used jdbc template.

File: components/org.wso2.finance.open.banking.conformance.mgt/src/main/java/org/wso2/finance/open/banking/conformance/mgt/dao/impl/TestPlanDAOImpl.java (hunk 12 above)
Comment (imesh94) [reply components/org.wso2.finance.open.banking.conformance.mgt/src/main/java/org/wso2/finance/open/banking/conformance/mgt/dao/impl/TestPlanDAOImpl.java]: done

File: samples/database/TableCreationQueries.txt [hunk 24]
```diff
@@ -0,0 +1,9 @@
+-------------------Create Table Queries------------------------
```
Comment (kaveenr) [inline samples/database/TableCreationQueries.txt]: Is this file necessary?

Comment (imesh94) [reply samples/database/TableCreationQueries.txt]: Added it to make things easy if we are creating a db script in future. (not necessary now since we are using an embedded db)

Comment (kaveenr) [reply samples/database/TableCreationQueries.txt]: Might as well make it a `.sql` file now itself.

File: db.properties [hunk 25]
```diff
@@ -0,0 +1,4 @@
+Dname=org.h2.Driver
```
Comment (kaveenr) [inline db.properties]: rename to`driverClassName` maybe

Comment (kaveenr) [reply db.properties]: Same for `uName` rename to 'userName`

File: samples/database/TableCreationQueries.txt (hunk 24 above)
Comment (kaveenr) [reply samples/database/TableCreationQueries.txt]: make this a '.sql' file instead

File: db.properties (hunk 25 above)
Comment (imesh94) [reply db.properties]: done
